src/ucp/rma/flush.c

```diff
@@ -684,55 +685,67 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_ep_flush, (ep), ucp_ep_h ep)
 {
-    ucp_context_t *context = worker->context;
-    ucp_request_t *request;
-    ucp_rsc_index_t rsc_index;
     ucp_worker_iface_t *wiface;
+    ucp_rsc_index_t rsc_index;
+    ucs_status_t status;
+
+    UCS_BITMAP_FOR_EACH_BIT(worker->context->tl_bitmap, rsc_index) {
+        wiface = ucp_worker_iface(worker, rsc_index);
+        if (wiface->iface == NULL) {
+            continue;
+        }
+
+        status = uct_iface_fence(wiface->iface, 0);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t ucp_worker_fence_strong(ucp_worker_h worker)
+{
+    ucp_request_t *request;
     ucs_status_ptr_t status_ptr;
     ucs_status_t status;
 
-    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
+    status_ptr = ucp_worker_flush_nbx_internal(worker,
+                                                &ucp_request_null_param,
+                                                UCT_FLUSH_FLAG_REMOTE);
+    if (ucs_unlikely(!UCS_PTR_IS_PTR(status_ptr))) {
+        return UCS_PTR_STATUS(status_ptr);
+    }
 
-    if (!context->config.worker_strong_fence) {
-        UCS_BITMAP_FOR_EACH_BIT(worker->context->tl_bitmap, rsc_index) {
-            wiface = ucp_worker_iface(worker, rsc_index);
-            if (wiface->iface == NULL) {
-                continue;
-            }
+    request = (ucp_request_t*)status_ptr - 1;
 
-            status = uct_iface_fence(wiface->iface, 0);
-            if (status != UCS_OK) {
-                goto out;
-            }
-        }
-        status = UCS_OK;
-    } else {
-        /* force using flush on EPs */
-        ucp_worker_flush_ops_count_add(worker, 1);
-        status_ptr = ucp_worker_flush_nbx_internal(worker,
-                                                   &ucp_request_null_param,
-                                                   UCT_FLUSH_FLAG_REMOTE);
-        
-        if (ucs_unlikely(!UCS_PTR_IS_PTR(status_ptr))) {
-            status = UCS_PTR_STATUS(status_ptr);
-            goto out_ops_count;
-        }
+    do {
+        ucp_worker_progress(worker);
+    } while (request->flush_worker.comp_count > 1);
+
+    ucp_worker_flush_complete_one(request, request->status, 1);
+    status = request->status;
+    ucp_request_put(request);
+    return status;
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_worker_fence, (worker), ucp_worker_h worker)
+{
+    ucs_status_t status;
 
-        request = (ucp_request_t*)status_ptr - 1;
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
 
-        do {
-            ucp_worker_progress(worker);
-        } while (request->flush_worker.comp_count > 1);
-        ucp_worker_flush_complete_one(request, request->status, 1);
-        status = request->status;
-        ucp_request_put(request);
-out_ops_count:
+    if (worker->context->config.worker_strong_fence) {
+        /* force using flush on EPs */
+        ucp_worker_flush_ops_count_add(worker, 1);
```

Comment:
IMO need to move ucp_worker_flush_ops_count_add inside ucp_worker_fence_strong

Reply:
it makes code bit more complex... ok, done